arrays/arrays.c

```diff
@@ -69,43 +94,82 @@ void resize_array(Array *arr) {
  *
  * Throw an error if the index is out of range.
  *****/
-char *arr_read(Array *arr, int index) {
+char *arr_read(Array *arr, int index)
+{
 
   // Throw an error if the index is greater than the current count
+  if (index >= arr->count)
+  {
+    printf("index out of range\n");
+
+    return NULL;
+  }
 
   // Otherwise, return the element at the given index
+  else
+  {
+    return arr->elements[index];
+  }
 }
 
-
 /*****
  * Insert an element to the array at the given index
  *****/
-void arr_insert(Array *arr, char *element, int index) {
+void arr_insert(Array *arr, char *element, int index)
+{
 
   // Throw an error if the index is greater than the current count
+  if (index > arr->count)
+  {
+    printf("index out of range\n");
+  }
 
-  // Resize the array if the number of elements is over capacity
+  else
+  {
+    // Resize the array if the number of elements is over capacity
+    if (arr->count == arr->capacity)
+    {
+      resize_array(arr);
+    }
 
-  // Move every element after the insert index to the right one position
+    // Move every element after the insert index to the right one position
+    for (int i = arr->count; i > index; i--)
+    {
+      arr->elements[i] = arr->elements[i - 1];
+    }
 
-  // Copy the element and add it to the array
+    // Copy the element and add it to the array
+    char *new_element = malloc(sizeof(char) * (strlen(element) + 1));
```

Comment:
Nice, this is exactly how it was intended for you to copy these over. Also you can consolidate lines 142 and 143 by simply calling `char *copy = strdup(element)` which does the malloc() and strcpy() all in one step.